src/module/elements.rs

```diff
@@ -0,0 +1,185 @@
+//! Table elements within a wasm module.
+
+use crate::error::Result;
+use crate::module::emit::{Emit, IdsToIndices};
+use crate::module::functions::FunctionId;
+use crate::module::tables::TableId;
+use crate::module::Module;
+use crate::passes::Used;
+use id_arena::{Arena, Id};
+use parity_wasm::elements;
+use std::collections::HashMap;
+
+/// A passive element segment identifier
+pub type ElementId = Id<Element>;
+
+/// A passive segment which contains a list of functions
+#[derive(Debug)]
+pub struct Element {
+    members: Vec<FunctionId>,
+}
+
+/// All element segments of a wasm module, used to initialize `anyfunc` tables,
+/// used as function pointers.
+#[derive(Debug, Default)]
+pub struct ModuleElements {
+    active: HashMap<TableId, Vec<Option<FunctionId>>>,
+    passive: Arena<Element>,
+    index_to_element_id: HashMap<u32, ElementId>,
+}
+
+impl ModuleElements {
+    /// Parses a raw was section into a fully-formed `ModuleElements` instance.
+    pub fn parse(module: &Module, section: &elements::ElementSection) -> Result<ModuleElements> {
+        let mut active = HashMap::new();
+        let mut passive = Arena::new();
+        let mut index_to_element_id = HashMap::new();
+
+        for (i, segment) in section.entries().iter().enumerate() {
+            if segment.passive() {
+                let mut list = Vec::with_capacity(segment.members().len());
+                for &func in segment.members() {
+                    match module.funcs.function_for_index(func) {
+                        Some(id) => list.push(id),
+                        None => failure::bail!("invalid segment initialization"),
```

Comment:
Maybe worth adding some `{}` containing the function index here for better debugging.